File: Playwright/PROJECTS/PROJECT_01_Login_Logout_Testing/pages/ProductsPage.ts
import { Page, Locator, expect } from '@playwright/test';
import { LoginPage } from './LoginPage';

/**
 * Page Object Pattern - Products Page (page after login)
 * Represents the page shown after successful login
 */
export class ProductsPage {
  readonly page: Page;
  readonly inventoryContainer: Locator;
  readonly menuButton: Locator;
  readonly logoutLink: Locator;
  readonly shoppingCart: Locator;
  readonly pageTitle: Locator;

  constructor(page: Page) {
    this.page = page;

    // Locators
    this.inventoryContainer = page.locator('#inventory_container');
    this.menuButton = page.locator('#react-burger-menu-btn');
    this.logoutLink = page.locator('#logout_sidebar_link');
    this.shoppingCart = page.locator('.shopping_cart_link');
    this.pageTitle = page.locator('.title');
  }

  /**
   * Check if the products page is loaded
   *
   * @returns true if page is loaded
   */
  async isLoaded(): Promise<boolean> {
    try {
      // Wait for URL to contain 'inventory' (products page)
      await this.page.waitForURL(/.*inventory.*/, { timeout: 15000 });
      
      // Wait for container element to be present and visible
      await expect(this.inventoryContainer).toBeVisible({ timeout: 10000 });
      
      return true;
    } catch (error) {
      // If first check fails, try alternative check
      try {
        const currentUrl = this.getCurrentUrl();
        if (currentUrl.includes('inventory')) {
          // If URL is correct, wait for menu which is always present
          await expect(this.menuButton).toBeVisible({ timeout: 5000 });
          return true;
        }
      } catch {
        return false;
      }
      return false;
    }
  }

  /**
   * Return the page title
   */
  async getPageTitle(): Promise<string | null> {
    try {
      return await this.pageTitle.textContent();
    } catch {
      return null;
    }
  }

  /**
   * Return the current URL
   */
  getCurrentUrl(): string {
    return this.page.url();
  }

  /**
   * Open the hamburger menu
   */
  async openMenu(): Promise<void> {
    await this.menuButton.click();
    // Wait for menu to open
    await expect(this.logoutLink).toBeVisible();
  }

  /**
   * Log out from the application
   *
   * @returns instance of login page
   */
  async logout(): Promise<LoginPage> {
    await this.openMenu();
    await this.logoutLink.click();
    
    // Wait for redirect to login page
    await this.page.waitForURL(/.*saucedemo\.com.*/, { timeout: 5000 });
    
    return new LoginPage(this.page);
  }

  /**
   * Check if the user is logged in
   *
   * @returns true if user is logged in (products page is visible)
   */
  async isLoggedIn(): Promise<boolean> {
    const isLoaded = await this.isLoaded();
    const currentUrl = this.getCurrentUrl();
    return isLoaded && currentUrl.includes('inventory');
  }
}

File: selenium-java-tests/PROJECT_01_Login_Logout_Testing/src/test/java/com/qa/automation/project01/pages/ProductsPage.java
package com.qa.automation.project01.pages;

import com.qa.automation.project01.config.TestConfig;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;

/**
 * Page Object for the Sauce Demo products (inventory) page shown after successful login.
 * <p>
 * Provides methods to verify that the products page is loaded, to read the page title and URL,
 * to open the hamburger menu and perform logout, and to check visibility of menu and cart.
 * All visibility and clickability checks use explicit waits (no fixed sleeps).
 */
public class ProductsPage {

    // ——— Locators (Sauce Demo inventory page) ———
    private static final By INVENTORY_CONTAINER = By.id("inventory_container");
    private static final By MENU_BUTTON = By.id("react-burger-menu-btn");
    private static final By LOGOUT_LINK = By.id("logout_sidebar_link");
    private static final By SHOPPING_CART_LINK = By.cssSelector(".shopping_cart_link");
    private static final By PAGE_TITLE = By.cssSelector(".title");
    /** Login button on the login page; used to confirm we are back on login after logout. */
    private static final By LOGIN_BUTTON = By.id("login-button");

    private final WebDriver driver;
    private final WebDriverWait wait;

    public ProductsPage(WebDriver driver) {
        this.driver = driver;
        this.wait = new WebDriverWait(driver, Duration.ofSeconds(TestConfig.getTimeoutSeconds()));
    }

    /**
     * Returns true when the products page is loaded: URL contains "inventory" and either
     * the inventory container or the menu button is visible. Used by tests to assert
     * successful login before checking title, menu, or performing logout.
     */
    public boolean isLoaded() {
        try {
            wait.until(ExpectedConditions.urlContains("inventory"));
            wait.until(ExpectedConditions.or(
                    ExpectedConditions.visibilityOfElementLocated(INVENTORY_CONTAINER),
                    ExpectedConditions.visibilityOfElementLocated(MENU_BUTTON)
            ));
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * Returns the main heading text (e.g. "Products") or null if the element is not found.
     * Used to assert the page title after login (TC-LOGIN-001).
     */
    public String getPageTitle() {
        try {
            WebElement title = driver.findElement(PAGE_TITLE);
            return title != null ? title.getText() : null;
        } catch (Exception e) {
            return null;
        }
    }

    /** Returns the current browser URL so tests can assert redirect (e.g. URL contains "inventory"). */
    public String getCurrentUrl() {
        return driver.getCurrentUrl();
    }

    /**
     * Opens the hamburger menu (top-left) and waits until the sidebar with the Logout link is visible.
     * Must be called before clicking the Logout link.
     */
    public void openMenu() {
        wait.until(ExpectedConditions.elementToBeClickable(MENU_BUTTON)).click();
        wait.until(ExpectedConditions.visibilityOfElementLocated(LOGOUT_LINK));
    }

    /**
     * Performs logout: opens the menu, clicks Logout, and waits until the login page is visible
     * (login button present). Waiting for the login button is more reliable than waiting for URL
     * change on some environments.
     *
     * @return a LoginPage instance for the same driver, so tests can continue assertions on the login page
     */
    public LoginPage logout() {
        openMenu();
        wait.until(ExpectedConditions.elementToBeClickable(LOGOUT_LINK)).click();
        wait.until(ExpectedConditions.visibilityOfElementLocated(LOGIN_BUTTON));
        return new LoginPage(driver);
    }

    /**
     * Returns true if the user is considered logged in: URL contains "inventory" and the page
     * is loaded (inventory or menu visible). Used to assert state before logout (TC-LOGOUT-001).
     */
    public boolean isLoggedIn() {
        return getCurrentUrl().contains("inventory") && isLoaded();
    }

    /** Returns the hamburger menu button element (e.g. for TC-LOGIN-004: assert menu is visible). */
    public WebElement getMenuButton() {
        return driver.findElement(MENU_BUTTON);
    }

    /** Returns the shopping cart link (e.g. for TC-LOGIN-004: assert cart icon is visible). */
    public WebElement getShoppingCartLink() {
        return driver.findElement(SHOPPING_CART_LINK);
    }
}
